### PR TITLE
Generic types and collections.

### PR DESCRIPTION
**Pull request by mckoss**
@tomlarkworthy 

This PR brings Bolt up to parity with all of the design patterns I've been using in JSON - and allows collections to be represented in schema (whereas before, they could only be referenced in path statements - which created some ugly non-DRY code).

Implemented are:
```
- Generic Types
- The Map<Key, Value> builtin generic type.
- Type "array-like" - Type[] === Map<String, Type>
- The key() function - used to validate the key of an object against some of it's contained data (a referential integrity rule).
```
See the language.md for documentation on these features of the language.

### CODE REVIEW

File: docs/language.md [hunk 1]
```diff
@@ -154,28 +225,52 @@ the `prior()` function:
...
 # Functions and Methods
 
 Functions must be simple return expressions with zero or more parameters.  All of the following
 examples are identical and can be used interchangably.
 
-    function myFunction(arg1, arg2) {
-      return arg1 == arg2.value;
-    }
+```javascript
+function myFunction(arg1, arg2) {
```
Comment (tomlarkworthy) [inline docs/language.md]: can we use a better name and a real example, maybe auth as thats the best use of functions

File: docs/language.md [hunk 2]
```diff
@@ -188,7 +283,7 @@ Rule expressions are a subset of JavaScript expressions, and include:
 
 These global variables are available in expressions:
 
-    root - The root of your Firebase database.
+    root - The root location of a Firebase database.
     auth - The current auth state (if auth != null the user is authenticated auth.uid
```
Comment (tomlarkworthy) [inline docs/language.md]: "The current auth state (if auth != null the user is authenticated auth.uid is their user identifier (a unique string value)."
>

"The current auth state (if auth != null, the user is authenticated, and auth.uid is their user identifier string)."

File: samples/generics.bolt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+  id: ProductID,
+  cost: Number
+}
+
+type ProductID extends String {
+  validate() = this.length <= 20;
+}
+
+type PushID extends String {
```
Comment (tomlarkworthy) [inline samples/generics.bolt]: great example!

File: src/ast.ts [hunk 4]
```diff
@@ -144,10 +160,26 @@ export function reference(base: Exp, prop: string | Exp): ExpReference {
 // immutability of the original expression).
 export function copyExp(exp: Exp): Exp {
   exp = <Exp> util.extend({}, exp);
-  if (exp.type === 'op' || exp.type === 'call') {
-    (<ExpOp> exp).args = util.copyArray((<ExpOp> exp).args);
+  switch (exp.type) {
+  case 'op':
```
Comment (tomlarkworthy) [inline src/ast.ts]: whats it return here?

File: src/ast.ts [hunk 5]
```diff
@@ -381,16 +406,17 @@ export class Symbols {
...
+    } else {
+      this[type][name] = object;
     }
-    this[type][name] = object;
+    return this[type][name];
   }
 
-  registerFunction(name: string, params: string[], body: Exp) {
-    this.register('functions', name, method(params, body));
+  registerFunction(name: string, params: string[], body: Exp): Method {
```
Comment (tomlarkworthy) [inline src/ast.ts]: name should be an enum really

Comment (tomlarkworthy) [inline src/ast.ts]: I wonder if register needs to be so broad in scope or whether registerFunction etc. for all the types would be better. I have not looked at its usage yet though.

File: src/ast.ts [hunk 6]
```diff
@@ -399,40 +425,68 @@ export class Symbols {
...
   }
 
-  registerSchema(name: string, derivedFrom: ExpType | ExpUnionType | void, properties, methods) {
-    methods = methods || {};
+  registerSchema(name: string,
+                 derivedFrom?: ExpType,
+                 properties?: { [prop: string]: ExpType },
+                 methods?: { [name: string]: Method },
+                 params?: string[]): Schema {
     properties = properties || {};
```
Comment (tomlarkworthy) [inline src/ast.ts]: You don't need these. Just make the parameter non-optional but give it a default value of properties = {}

Comment (tomlarkworthy) [inline src/ast.ts]: Oh I see derivedFrom is blocking it a bit.

File: src/ast.ts [hunk 7]
```diff
@@ -399,40 +425,68 @@ export class Symbols {
...
+      methods: methods,
     };
-    this.register('schema', name, s);
-  }
-
-  isDerivedFrom(descendant: string, ancestor: string, visited?: { [name: string]: boolean}) {
-    if (!visited) {
-      visited = {};
+    // Only add params if this is a Generic schema.
+    if (params) {
```
Comment (tomlarkworthy) [inline src/ast.ts]: Can we get rid of special casing generics by making params default to [], and have a generic validator that doesn't change anything?

File: src/ast.ts [hunk 8]
```diff
@@ -399,40 +425,68 @@ export class Symbols {
...
-    if (descendant === ancestor) {
+  isDerivedFrom(type: ExpType, ancestor: string): boolean {
+    if (ancestor === 'Any') {
       return true;
     }
 
-    var schema = this.schema[descendant];
-    if (!schema) {
-      return false;
+    switch (type.type) {
```
Comment (tomlarkworthy) [inline src/ast.ts]: these two *almost* identical switch statements in a row, but one returns in the middle for a certain condition read badly. Can we think of a better way of expressing isDerivedFrom that doesn't have quite so many non-linear control flow kick outs

File: src/rules-generator.ts [hunk 9]
```diff
@@ -239,15 +226,61 @@ export class Generator {
...
+    let map = this.symbols.registerSchema('Map', ast.typeType('Any'), undefined, undefined,
+                                          ['Key', 'Value']);
+    map.getValidator = this.getMapValidator.bind(this);
+  }
+
+  // type Map<Key, Value> => {
+  //   $key: {
+  //     '.validate': $key instanceof Key and this instanceof Value;
+  // }
+  // Key must derive from String
```
Comment (tomlarkworthy) [inline src/rules-generator.ts]: move this comment to above the if (keyType) line

File: src/rules-generator.ts [hunk 10]
```diff
@@ -239,15 +226,61 @@ export class Generator {
...
+  getMapValidator(params: ast.Exp[]): Validator {
+    let keyType = <ast.ExpSimpleType> params['Key'];
+    let valueType = <ast.ExpType> params['Value'];
+    if (keyType.type !== 'type' || !this.symbols.isDerivedFrom(keyType, 'String')) {
```
Comment (tomlarkworthy) [inline src/rules-generator.ts]: here

File: src/rules-generator.ts [hunk 11]
```diff
@@ -239,15 +226,61 @@ export class Generator {
...
+      throw new Error(errors.invalidMapKey + "  (" + decodeExpression(keyType) + " does not)");
+    }
+
+    let validator = <Validator> {};
```
Comment (tomlarkworthy) [inline src/rules-generator.ts]: I am quite confused over Validator and extendValidator. The comment on Validator doesn't help me figure out what we are doing here. Is extend doing a boolean & ?

File: src/ast.ts (hunk 4 above)
Comment (mckoss) [inline src/ast.ts]: 'op' and 'call' have the same structure - they both need to copy their args (array) property.

File: src/ast.ts (hunk 8 above)
Comment (mckoss) [inline src/ast.ts]: Good point - this was a refactoring/inlining for a function ... I'll clean this up.

File: src/rules-generator.ts [hunk 12]
```diff
@@ -255,37 +288,170 @@ export class Generator {
   //     ".validate": [<expression>, ...]
   // All expressions will be ANDed together to form the file expresion.
   // Intermediate nodes can be "prop" or "$prop" values.
-  createValidator(schemaName: string): Validator {
+  createValidator(type: ast.ExpType): Validator {
```
Comment (tomlarkworthy) [inline src/rules-generator.ts]: checkpoint for review ... to be continued

File: docs/language.md [hunk 13]
```diff
@@ -48,41 +52,106 @@ a `validate` expression.
...
+Any place a Type can be used, it can be replaced with a Type expression.  We support
+three types of type expressions:
+
+### Union Types
 
     Type1 | Type2    - Value can be either of two types.
     Type | Null      - An optional `Type` value (value can be deleted or missing).
 
+
+### Map Types (Collections)
```
Comment (matjaz) [inline docs/language.md]: is there a way to check collection length?
```
type Group
  members: User[]

  validate() = this.members.length > 0
```

File: src/ast.ts (hunk 4 above)
Comment (matjaz) [inline src/ast.ts]: http://jshint.com/docs/

search `/* falls through */`

**Comment by matjaz** [comment]
nice job!

File: src/rules-generator.ts (hunk 11 above)
Comment (mckoss) [inline src/rules-generator.ts]: Added comment to validator structure at top of file.

File: src/rules-generator.ts (hunk 9 above)
Comment (mckoss) [inline src/rules-generator.ts]: This was meant to be an over-arching method description.

File: src/ast.ts (hunk 4 above)
Comment (mckoss) [inline src/ast.ts]: I just checked - I do have that warning turned on - it does not file in this particular case of am empty case block "falling through" to a non-empty one.

http://eslint.org/docs/rules/no-fallthrough.html

File: docs/language.md (hunk 13 above)
Comment (mckoss) [inline docs/language.md]: Not really.  Collections are just Objects with unique keys for each member.

If you want to limit the max size for small collections you could do something like this (not that all keys are strings - not integers):
```
type SingleDigit extends String {
  validate() = this.length == 1 && this >= '0' && this <= '9';
}

type Group {
  members: Map<SingleDigit, User>
}
```

See this sample:

https://github.com/firebase/bolt/blob/koss-generics-2/samples/groups.bolt

File: src/ast.ts (hunk 5 above)
Comment (mckoss) [inline src/ast.ts]: register function is just a helper to consolidate error reporting in one place while updated the various parts of the symbol table.

File: src/ast.ts (hunk 6 above)
Comment (mckoss) [inline src/ast.ts]: done

File: src/ast.ts (hunk 7 above)
Comment (mckoss) [inline src/ast.ts]: Good idea - done.

File: src/ast.ts (hunk 8 above)
Comment (mckoss) [inline src/ast.ts]: done

File: docs/language.md (hunk 1 above)
Comment (mckoss) [inline docs/language.md]: done

File: docs/language.md (hunk 2 above)
Comment (mckoss) [inline docs/language.md]: reworded

**Pull request merged**